instrument-starters/opentracing-spring-cloud-core/src/main/java/io/opentracing/contrib/spring/cloud/async/instrument/TracedListenableFutureCallback.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2017-2020 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.contrib.spring.cloud.async.instrument;
+
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.concurrent.FailureCallback;
+import org.springframework.util.concurrent.ListenableFutureCallback;
+import org.springframework.util.concurrent.SuccessCallback;
+
+/**
+ * @author MiNG
+ */
+public class TracedListenableFutureCallback<T> implements ListenableFutureCallback<T> {
+
+  private final SuccessCallback<T> successDelegate;
+  private final FailureCallback failureDelegate;
+  private final Tracer tracer;
+  private final Span span;
+
+  public TracedListenableFutureCallback(ListenableFutureCallback<T> delegate, Tracer tracer) {
+    this(delegate, delegate, tracer, tracer.activeSpan());
```

Comment:
Is this correct? Should be the success and failure callback the same instance?

Reply:
Yes, it's ok to put them together.
Actually, the separated version is added since spring 4.1 optimize for lambda in java8.
And the null handling behavior follows spring-core, which is raising exceptions instantly.